idb/postgres/internal/writer/writer.go

```diff
@@ -278,17 +316,37 @@ func getTransactionParticipants(stxnad transactions.SignedTxnWithAD) []basics.Ad
 	return res
 }
 
-func addTransactionParticipation(block *bookkeeping.Block, batch *pgx.Batch) error {
-	for i, stxnad := range block.Payset {
-		// TODO: replace with a function from go-algorand.
-		participants := getTransactionParticipants(stxnad.SignedTxnWithAD)
+// addInnerTransactionParticipation traverses the inner transaction tree and
+// adds txn participation records for each. It performs a preorder traversal
+// to correctly compute the intra round offset, the offset for the next
+// transaction is returned.
+func addInnerTransactionParticipation(stxnad transactions.SignedTxnWithAD, round, intra uint64, batch *pgx.Batch) uint64 {
```

Comment:
I think this function can be modified to also add participation for the outer txn and reduce code duplication with `addTransactionParticipation()`.

Reply:
We still need the `Payset` loop and the `InnerTxns` loop. Maybe there's an elegant solution, but I think this one is OK.